File: packages/raisin-client-js/src/branches.ts
/**
 * Branch management operations
 */

import { RequestContext, RequestType } from './protocol';

/**
 * Branch divergence information (similar to "N commits ahead/behind").
 */
export interface BranchDivergence {
  /** Number of commits on the compared branch not on the base branch */
  ahead: number;
  /** Number of commits on the base branch not on the compared branch */
  behind: number;
  /** Common ancestor revision (HLC string, e.g. "1705584000000-5") */
  common_ancestor: string;
}

/**
 * Per-node change in a branch diff (relative to the merge-base).
 */
export interface NodeDiffInfo {
  /** Node id that changed */
  node_id: string;
  /** Workspace the node belongs to */
  workspace: string;
  /** Path on the diffed branch (null when only a tombstone exists, e.g. some deletes) */
  path: string | null;
  /** Operation: "added" | "modified" | "deleted" | "reordered" */
  operation: string;
  /** Translation locale when the change is to a translation (absent = base node) */
  translation_locale?: string;
}

/**
 * Per-node diff of a branch relative to a base branch's merge-base.
 *
 * Unlike {@link BranchDivergence} (ahead/behind counts only), this
 * enumerates exactly which nodes changed since the two branches diverged.
 */
export interface BranchDiff {
  /** The common ancestor (merge-base) the diff is computed against (HLC string) */
  common_ancestor: string;
  /** Nodes present on the branch but not on the base */
  added: NodeDiffInfo[];
  /** Nodes changed on the branch (operation "modified" / "reordered") */
  modified: NodeDiffInfo[];
  /** Nodes removed on the branch */
  deleted: NodeDiffInfo[];
}

/**
 * Per-node change from a cross-branch node-set copy.
 */
export interface CrossBranchNodeChange {
  /** Node id (ids are preserved across branches) */
  node_id: string;
  /** Path of the node on the target branch */
  path: string;
  /** NodeType of the node */
  node_type: string;
  /** Operation on the target branch: "added" | "modified" | "deleted" */
  operation: string;
}

/**
 * Summary of a cross-branch node-set copy (see {@link Branches.copyNodes}).
 */
export interface CrossBranchCopySummary {
  /** Number of nodes written to the target branch (added + modified) */
  copied: number;
  /** Number of target-branch nodes pruned (deleteMissing only) */
  deleted: number;
  /** The single revision all changes were committed under (HLC string) */
  revision: string;
  /** Per-node change list */
  changes: CrossBranchNodeChange[];
}

export class Branches {
  private sendRequest: (payload: unknown, requestType: string) => Promise<unknown>;

  constructor(
    _context: RequestContext,
    sendRequest: (payload: unknown, requestType: string) => Promise<unknown>
  ) {
    this.sendRequest = sendRequest;
  }

  /**
   * Create a new branch
   */
  async create(
    name: string,
    options?: {
      fromRevision?: string;
      fromBranch?: string;
    }
  ): Promise<unknown> {
    return this.sendRequest(
      {
        name,
        from_revision: options?.fromRevision,
        from_branch: options?.fromBranch
      },
      RequestType.BranchCreate
    );
  }

  /**
   * Get a branch by name
   */
  async get(name: string): Promise<unknown> {
    return this.sendRequest(
      {
        name
      },
      RequestType.BranchGet
    );
  }

  /**
   * List all branches
   */
  async list(): Promise<unknown[]> {
    const result = await this.sendRequest(
      {},
      RequestType.BranchList
    );
    return Array.isArray(result) ? result : [];
  }

  /**
   * Delete a branch
   */
  async delete(name: string): Promise<unknown> {
    return this.sendRequest(
      {
        name
      },
      RequestType.BranchDelete
    );
  }

  /**
   * Get the HEAD revision of a branch
   */
  async getHead(name: string): Promise<unknown> {
    return this.sendRequest(
      {
        name
      },
      RequestType.BranchGetHead
    );
  }

  /**
   * Update the HEAD revision of a branch
   */
  async updateHead(name: string, revision: string): Promise<unknown> {
    return this.sendRequest(
      {
        name,
        revision
      },
      RequestType.BranchUpdateHead
    );
  }

  /**
   * Merge a source branch into a target branch
   */
  async merge(
    sourceBranch: string,
    targetBranch: string,
    options?: {
      strategy?: string;
      message?: string;
    }
  ): Promise<unknown> {
    return this.sendRequest(
      {
        source_branch: sourceBranch,
        target_branch: targetBranch,
        strategy: options?.strategy,
        message: options?.message
      },
      RequestType.BranchMerge
    );
  }

  /**
   * Compare two branches to calculate divergence
   */
  async compare(branch: string, baseBranch: string): Promise<BranchDivergence> {
    return this.sendRequest(
      {
        branch,
        base_branch: baseBranch
      },
      RequestType.BranchCompare
    ) as Promise<BranchDivergence>;
  }

  /**
   * Per-node diff of `branch` relative to `baseBranch` (added/modified/deleted).
   *
   * Cheaper than reading both branch listings: cost scales with the number of
   * changed nodes since the branches diverged, not the total content.
   */
  async diff(branch: string, baseBranch: string): Promise<BranchDiff> {
    return this.sendRequest(
      {
        branch,
        base_branch: baseBranch
      },
      RequestType.BranchDiff
    ) as Promise<BranchDiff>;
  }

  /**
   * Copy a node set from `sourceBranch` onto `targetBranch` (branch
   * promotion): node ids are preserved, so repeated copies update the same
   * target nodes; all changes land in ONE atomic commit on the target branch.
   *
   * Each root's parent path must already exist on the target branch. With
   * `deleteMissing`, target nodes under the roots that are absent from the
   * copied source set are pruned (one-way sync).
   *
   * `sourceRevision` pins the read to a point in time. A promotion of a large
   * set is not instantaneous, so without it the copy reads each node when its
   * turn comes and a writer working in parallel lands partly inside the result
   * — a torn snapshot. Pass the revision captured when the operation was
   * decided (see `getHead`) to copy exactly that state. The target branch is
   * always resolved at its head: the pin says WHAT to copy, never where.
   */
  async copyNodes(
    sourceBranch: string,
    targetBranch: string,
    opts: {
      workspace: string;
      roots: string[];
      recursive?: boolean;
      deleteMissing?: boolean;
      sourceRevision?: string;
    }
  ): Promise<CrossBranchCopySummary> {
    return this.sendRequest(
      {
        source_branch: sourceBranch,
        target_branch: targetBranch,
        workspace: opts.workspace,
        roots: opts.roots,
        recursive: opts.recursive ?? true,
        delete_missing: opts.deleteMissing ?? false,
        source_revision: opts.sourceRevision ?? null
      },
      RequestType.BranchCopyNodes
    ) as Promise<CrossBranchCopySummary>;
  }
}

File: crates/raisin-functions/src/runtime/bindings/methods/branches.rs
// SPDX-License-Identifier: BSL-1.1
//
// RaisinDB - Git-like hierarchical multi model database
// Copyright (C) 2019-2025 SOLUTAS GmbH, Switzerland

//! Branch API bindings (`raisin.branches.*`)

use crate::api::FunctionApi;
use crate::runtime::bindings::registry::{
    ApiMethodDescriptor, ArgParser, ArgSpec, ArgType, InvokeResult, ReturnType,
};
use futures::future::BoxFuture;
use raisin_error::Result;
use serde_json::Value;
use std::sync::Arc;

/// Get all branch method descriptors
pub fn methods() -> Vec<ApiMethodDescriptor> {
    vec![
        // branches.diff(branch, baseBranch) - per-node diff of `branch`
        // relative to `baseBranch`'s merge-base. Returns
        // { common_ancestor, added: [...], modified: [...], deleted: [...] }.
        ApiMethodDescriptor {
            internal_name: "branches_diff",
            js_name: "diff",
            py_name: "diff",
            category: "branches",
            args: vec![
                ArgSpec::new("branch", ArgType::String),
                ArgSpec::new("baseBranch", ArgType::String),
            ],
            return_type: ReturnType::Json,
            invoker: |api: Arc<dyn FunctionApi>,
                      args: Vec<Value>|
             -> BoxFuture<'static, Result<InvokeResult>> {
                Box::pin(async move {
                    let mut parser = ArgParser::new(&args);
                    let branch = parser.string()?;
                    let base_branch = parser.string()?;

                    let result = api.branch_diff(&branch, &base_branch).await?;
                    Ok(InvokeResult::Json(result))
                })
            },
        },
        // branches.compare(branch, baseBranch) - divergence (commits
        // ahead/behind) of `branch` relative to `baseBranch`. Returns
        // { ahead, behind, common_ancestor }.
        ApiMethodDescriptor {
            internal_name: "branches_compare",
            js_name: "compare",
            py_name: "compare",
            category: "branches",
            args: vec![
                ArgSpec::new("branch", ArgType::String),
                ArgSpec::new("baseBranch", ArgType::String),
            ],
            return_type: ReturnType::Json,
            invoker: |api: Arc<dyn FunctionApi>,
                      args: Vec<Value>|
             -> BoxFuture<'static, Result<InvokeResult>> {
                Box::pin(async move {
                    let mut parser = ArgParser::new(&args);
                    let branch = parser.string()?;
                    let base_branch = parser.string()?;

                    let result = api.branch_compare(&branch, &base_branch).await?;
                    Ok(InvokeResult::Json(result))
                })
            },
        },
        // branches.copyNodes(sourceBranch, targetBranch, opts) - copy a node
        // set across branches (ids preserved, one atomic commit on the
        // target). opts: { workspace, roots, recursive?, deleteMissing? }.
        // Returns { copied, deleted, revision, changes: [...] }.
        ApiMethodDescriptor {
            internal_name: "branches_copyNodes",
            js_name: "copyNodes",
            py_name: "copy_nodes",
            category: "branches",
            args: vec![
                ArgSpec::new("sourceBranch", ArgType::String),
                ArgSpec::new("targetBranch", ArgType::String),
                ArgSpec::new("opts", ArgType::Json),
            ],
            return_type: ReturnType::Json,
            invoker: |api: Arc<dyn FunctionApi>,
                      args: Vec<Value>|
             -> BoxFuture<'static, Result<InvokeResult>> {
                Box::pin(async move {
                    let mut parser = ArgParser::new(&args);
                    let source_branch = parser.string()?;
                    let target_branch = parser.string()?;
                    let opts = parser.json()?;

                    let result = api
                        .branch_copy_nodes(&source_branch, &target_branch, opts)
                        .await?;
                    Ok(InvokeResult::Json(result))
                })
            },
        },
    ]
}
